README.md
# Invoke-FslShrinkDisk.ps1

[![Build Status](https://dev.azure.com/jimoyle/Invoke-FslShrinkDisk/_apis/build/status/FSLogix.Invoke-FslShrinkDisk?branchName=master)](https://dev.azure.com/jimoyle/Invoke-FslShrinkDisk/_build/latest?definitionId=1&branchName=master)

[![Codacy Badge](https://app.codacy.com/project/badge/Grade/725c8d2481044524b331d3b207971ddf)](https://www.codacy.com/gh/FSLogix/Invoke-FslShrinkDisk?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=FSLogix/Invoke-FslShrinkDisk&amp;utm_campaign=Badge_Grade)

## .SYNOPSIS

Shrinks FSLogix Profile and O365 dynamically expanding disk(s).

## .DESCRIPTION

FSLogix profile and O365 virtual hard disks are in the vhd or vhdx file format. By default, the disks created will be in Dynamically Expanding format rather than Fixed format.  This script does not support reducing the size of a Fixed file format.

Dynamically Expanding disks do not natively shrink when the volume of data within them reduces, they stay at the 'High water mark' of historical data volume within them.

This means that Enterprises can wish to reclaim whitespace inside the disks to keep cost down if the storage is cloud based, or make sure they don’t exceed capacity limits if storage is on-premises.

This Script is designed to work at Enterprise scale to reduce the size of thousands of disks in the shortest time possible.
This script can be run from any machine in your environment it does not need to be run from a file server hosting the disks.  It does not need the Hyper-V role installed.

Powershell version 5.x and 7 and above are supported for this script. It needs to be run as administrator due to the requirement for mounting disks to the OS where the script is run.

This tool is multi-threaded and will take advantage of multiple CPU cores on the machine from which you run the script.  It is not advised to run more than 2x the threads of your available cores on your machine.  You could also use the number of threads to throttle the load on your storage.

Reducing the size of a virtual hard disk is a storage intensive activity.  The activity is more in file system metadata operations than pure IOPS, so make sure your storage controllers can handle the load.  The storage load occurs on the location where the disks are stored not on the machine where the script is run from.   I advise running the script out of hours if possible, to avoid impacting other users on the storage.

With the intention of reducing the storage load to the minimum possible, you can configure the script to only shrink the disks where you will see the most benefit.  You can delete disks which have not been accessed in x number of days previously (configurable).  Deletion of disks is not enabled by default.

By default the script will not run on any disk with less than 5% whitespace inside (configurable).  The script can optionally also not run on disks smaller than (x)GB (configurable) as it’s possible that even a large % of whitespace in small disks won’t result in a large capacity reclamation, but even shrinking a small amount of capacity will cause storage load.
The script will output a csv in the following format:

    "Name","DiskState","OriginalSizeGB","FinalSizeGB","SpaceSavedGB","FullName"
    "Profile_user1.vhdx","Success","4.35","3.22","1.13",\\Server\Share\ Profile_user1.vhdx "
    "Profile_user2.vhdx","Success","4.75","3.12","1.63",\\Server\Share\ Profile_user2.vhdx

### Possible Information values for DiskState are as follows

| DiskState | Meaning |
|-----|-----|
| Success                  | Disk has been successfully processed and shrunk |
| Ignored                  | Disk was less than the size configured in -IgnoreLessThanGB parameter |
| Deleted                  | Disk was last accessed before the number of days configured in the -DeleteOlderThanDays parameter and was successfully deleted |
| DiskLocked                | Disk could not be mounted due to being in use |
| LessThan(x)%FreeInsideDisk| Disk contained less whitespace than configured in -RatioFreeSpace parameter and was ignored for processing |

### Possible Error values for DiskState are as follows

| DiskState | Meaning |
|-----|-----|
| FileIsNotDiskFormat           | Disk file extension was not vhd or vhdx  |
| DiskDeletionFailed            | Disk was last accessed before the number of days configured in the -DeleteOlderThanDays parameter and was not successfully deleted |
| NoPartitionInfo               | Could not get partition information for partition 1 from the disk |
| PartitionShrinkFailed         | Failed to Shrink partition as part of the disk processing |
| DiskShrinkFailed              | Could not shrink Disk |
| PartitionSizeRestoreFailed    | Failed to Restore partition as part of the disk processing |

If the diskstate shows an error value from the list above, manual intervention may be required to make the disk usable again.

If you inspect your environment you will probably see that there are a few disks that are consuming a lot of capacity targeting these by using the minimum disk size configuration would be a good step.  To grab a list of disks and their sizes from a share you could use this oneliner by replacing < yourshare > with the path to the share containing the disks.

    Get-ChildItem -Path <yourshare> -Filter "*.vhd*" -Recurse -File | Select-Object Name, @{n = 'SizeInGB'; e = {[math]::round($_.length/1GB,2)}}

All this oneliner does is gather the names and sizes of the virtual hard disks from your share.  To export this information to a file readable by excel, use the following replacing both < yourshare > and < yourcsvfile.csv >.  You can then open the csv file in excel.

    Get-ChildItem -Path <yourshare> -Filter "*.vhd*" -Recurse -File | Select-Object Name, @{n = 'SizeInGB'; e = {[math]::round($_.length/1GB,2)}} | Export-Csv -Path < yourcsvfile.csv >

## .NOTES
Whilst I work for Microsoft and used to work for FSLogix, this is not officially released software from either company.  This is purely a personal project designed to help the community.  If you require support for this tool please raise an issue on the GitHub repository linked below

## .PARAMETER Path
The path to the folder/share containing the disks. You can also directly specify a single disk. UNC paths are supported.

## .PARAMETER Recurse
Gets the disks in the specified locations and in all child items of the locations

## .PARAMETER IgnoreLessThanGB
The disk size in GB under which the script will not process the file.

## .PARAMETER DeleteOlderThanDays
If a disk ‘last access time’ is older than todays date minus this value, the disk will be deleted from the share.  This is a permanent action.

## .PARAMETER LogFilePath
All disk actions will be saved in a csv file for admin reference.  The default location for this csv file is the user’s temp directory.  The default filename is in the following format: FslShrinkDisk 2020-04-14 19-36-19.csv

## .PARAMETER PassThru
Returns an object representing the item with which you are working. By default, this cmdlet does not generate any pipeline output.

### .PARAMETER ThrottleLimit
Specifies the number of disks that will be processed at a time. Further disks in the queue will wait till a previous disk has finished up to a maximum of the ThrottleLimit.  The  default value is 8.

## .PARAMETER RatioFreeSpace
The minimum percentage of white space in the disk before processing will start as a decimal between 0 and 1 eg 0.2 is 20% 0.65 is 65%. The Default is 0.05. This means that if the available size reduction is less than 5%, then no action will be taken.  To try and shrink all files no matter how little the gain set this to 0.

## .INPUTS
You can pipe the path into the command which is recognised by type, you can also pipe any parameter by name. It will also take the path positionally

## .OUTPUTS
This script outputs a csv file with the result of the disk processing.  It will optionally produce a custom object with the same information

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path c:\Profile_user1.vhdx
This shrinks a single disk on the local file system

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path \\server\share -Recurse
This shrinks all disks in the specified share recursively

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path \\server\share -Recurse -IgnoreLessThanGB 3
This shrinks all disks in the specified share recursively, except for files under 3GB in size which it ignores.

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path \\server\share -Recurse -DeleteOlderThanDays 90
This shrinks all disks in the specified share recursively and deletes disks which were not accessed within the last 90 days.

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path \\server\share -Recurse -LogFilePath C:\MyLogFile.csv
This shrinks all disks in the specified share recursively and changes the default log file location to C:\MyLogFile.csv

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path \\server\share -Recurse -PassThru

    Name:			Profile_user1.vhdx
    DiskState:		Success
    OriginalSizeGB:		4.35
    FinalSizeGB:		3.22
    SpaceSavedGB:		1.13
    FullName:		\\Server\Share\ Profile_user1.vhdx

This shrinks all disks in the specified share recursively and passes the result of the disk processing to the pipeline as an object as well as saving the results in a csv in the default location.

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path \\server\share -Recurse -ThrottleLimit 20
This shrinks all disks in the specified share recursively increasing the number of threads used to 20 from the default 8.

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path \\server\share -Recurse -RatioFreeSpace 0.3
This shrinks all disks in the specified share recursively while not processing disks which have less than 30% whitespace instead of the default 15%.

## .EXAMPLE
    C:\PS> Invoke-FslShrinkDisk.ps1 -Path \\server\share -Recurse -PassThru IgnoreLessThanGB 3 -DeleteOlderThanDays 90 -LogFilePath C:\MyLogFile.csv -ThrottleLimit 20 -RatioFreeSpace 0.3
This does all of the above examples, but together.

## .LINK
<https://github.com/FSLogix/Invoke-FslShrinkDisk/>


Changelog.md
# ChangeLog

## 2020 09 28

- Many improvements to cope with the underlying disk subsystem returning incomplete or non-existent objects when under stress
- New checking and remediation for pre-requisite services
- Introduced throttling of threads when the script is being run on too few cores for the threads specified
- Introduced better error messages
- Added better parameter validation

## 2020 08 12

- Removed unecessary partition resize from script
- Fixed Diskpart wasn't running against disks with spaces in the path
- Improved Shrink for legacy vhd disks making them more consistent
- Added better error reporting if the disk doesn't mount properly
- Improved file filtering so it doesn't pick up false positives

Functions/Private/Shrink-OneDisk.ps1
function Shrink-OneDisk {
    [CmdletBinding()]

    Param (
        [Parameter(
            ValuefromPipelineByPropertyName = $true,
            ValuefromPipeline = $true,
            Mandatory = $true
        )]
        [System.IO.FileInfo]$Disk,

        [Parameter(
            ValuefromPipelineByPropertyName = $true
        )]
        [Int]$DeleteOlderThanDays,

        [Parameter(
            ValuefromPipelineByPropertyName = $true
        )]
        [Int]$IgnoreLessThanGB,

        [Parameter(
            ValuefromPipelineByPropertyName = $true
        )]
        [double]$RatioFreeSpace = 0.05,

        [Parameter(
            ValuefromPipelineByPropertyName = $true
        )]
        [string]$LogFilePath = "$env:TEMP\FslShrinkDisk $(Get-Date -Format yyyy-MM-dd` HH-mm-ss).csv",

        [Parameter(
            ValuefromPipelineByPropertyName = $true
        )]
        [switch]$Passthru

    )

    BEGIN {
        #Requires -RunAsAdministrator
        Set-StrictMode -Version Latest
        $hyperv = $false
    } # Begin
    PROCESS {
        #Grab size of disk being porcessed
        $originalSizeGB = [math]::Round( $Disk.Length / 1GB, 2 )

        #Set default parameter values for the Write-VhdOutput command to prevent repeating code below, these can be overridden as I need to.
        $PSDefaultParameterValues = @{
            "Write-VhdOutput:Path"           = $LogFilePath
            "Write-VhdOutput:Name"           = $Disk.Name
            "Write-VhdOutput:DiskState"      = $null
            "Write-VhdOutput:OriginalSizeGB" = $originalSizeGB
            "Write-VhdOutput:FinalSizeGB"    = $originalSizeGB
            "Write-VhdOutput:SpaceSavedGB"   = 0
            "Write-VhdOutput:FullName"       = $Disk.FullName
            "Write-VhdOutput:Passthru"       = $Passthru
        }

        #Check it is a disk
        if ($Disk.Extension -ne '.vhd' -and $Disk.Extension -ne '.vhdx' ) {
            Write-VhdOutput -DiskState 'FileIsNotDiskFormat'
            return
        }

        #If it's older than x days delete disk
        If ( $DeleteOlderThanDays ) {
            #Last Access time isn't always reliable if diff disks are used so lets be safe and use the most recent of access and write
            $mostRecent = $Disk.LastAccessTime, $Disk.LastWriteTime | Measure-Object -Maximum | Select-Object -ExpandProperty Maximum
            if ($mostRecent -lt (Get-Date).AddDays(-$DeleteOlderThanDays) ) {
                try {
                    Remove-Item $Disk.FullName -ErrorAction Stop -Force
                    Write-VhdOutput -DiskState "Deleted" -FinalSizeGB 0 -SpaceSavedGB $originalSizeGB
                }
                catch {
                    Write-VhdOutput -DiskState 'DiskDeletionFailed'
                }
                return
            }
        }

        #As disks take time to process, if you have a lot of disks, it may not be worth shrinking the small ones
        if ( $IgnoreLessThanGB -and $originalSizeGB -lt $IgnoreLessThanGB ) {
            Write-VhdOutput -DiskState 'Ignored'
            return
        }

        #Initial disk Mount
        try {
            $mount = Mount-FslDisk -Path $Disk.FullName -PassThru -ErrorAction Stop
        }
        catch {
            $diskError = $error[0]
            Write-VhdOutput -DiskState $diskError.exception.message
            return
        }

        $partInfo = Get-Partition -DiskNumber $mount.DiskNumber | Where-Object -Property 'Type' -EQ -Value 'Basic'
        Get-Volume -Partition $partInfo | Optimize-Volume

        #Grab partition information so we know what size to shrink the partition to and what to re-enlarge it to.  This helps optimise-vhd work at it's best
        try {
            $partitionsize = Get-PartitionSupportedSize -InputObject $partInfo -ErrorAction Stop
            $sizeMax = $partitionsize.SizeMax
        }
        catch {
            Write-VhdOutput -DiskState 'NoPartitionInfo'
            $mount | DisMount-FslDisk
            return
        }


        #If you can't shrink the partition much, you can't reclaim a lot of space, so skipping if it's not worth it. Otherwise shink partition and dismount disk

        if ( $partitionsize.SizeMin -gt $disk.Length ) {
            Write-VhdOutput -DiskState "SkippedAlreadyMinimum"
            $mount | DisMount-FslDisk
            return
        }


        if (($partitionsize.SizeMin / $disk.Length) -gt (1 - $RatioFreeSpace) ) {
            Write-VhdOutput -DiskState "LessThan$(100*$RatioFreeSpace)%FreeInsideDisk"
            $mount | DisMount-FslDisk
            return
        }

        #If I decide to add Hyper-V module support, I'll need this code later
        if ($hyperv -eq $true) {

            #In some cases you can't do the partition shrink to the min so increasing by 100 MB each time till it shrinks
            $i = 0
            $resize = $false
            $targetSize = $partitionsize.SizeMin
            $sizeBytesIncrement = 100 * 1024 * 1024

            while ($i -le 5 -and $resize -eq $false) {

                try {
                    Resize-Partition -InputObject $partInfo -Size $targetSize -ErrorAction Stop
                    $resize = $true
                }
                catch {
                    $resize = $false
                    $targetSize = $targetSize + $sizeBytesIncrement
                    $i++
                }
                finally {
                    Start-Sleep 1
                }
            }

            #Whatever happens now we need to dismount

            if ($resize -eq $false) {
                Write-VhdOutput -DiskState "PartitionShrinkFailed"
                $mount | DisMount-FslDisk
                return
            }
        }

        $mount | DisMount-FslDisk

        #Change the disk size and grab the new size

        $retries = 0
        $success = $false
        #Diskpart is a little erratic and can fail occasionally, so stuck it in a loop.
        while ($retries -lt 30 -and $success -ne $true) {

            $tempFileName = "$env:TEMP\FslDiskPart$($Disk.Name).txt"

            #Let's put diskpart into a function just so I can use Pester to Mock it
            function invoke-diskpart ($Path) {
                #diskpart needs you to write a txt file so you can automate it, because apparently it's 1989.
                #A better way would be to use optimize-vhd from the Hyper-V module,
                #   but that only comes along with installing the actual role, which needs CPU virtualisation extensions present,
                #   which is a PITA in cloud and virtualised environments where you can't do Hyper-V.
                #MaybeDo, use hyper-V module if it's there if not use diskpart? two code paths to do the same thing probably not smart though
                Set-Content -Path $Path -Value "SELECT VDISK FILE=`'$($Disk.FullName)`'"
                Add-Content -Path $Path -Value 'attach vdisk readonly'
                Add-Content -Path $Path -Value 'COMPACT VDISK'
                Add-Content -Path $Path -Value 'detach vdisk'
                $result = DISKPART /s $Path
                Write-Output $result
            }

            $diskPartResult = invoke-diskpart -Path $tempFileName

            #diskpart doesn't return an object (1989 remember) so we have to parse the text output.
            if ($diskPartResult -contains 'DiskPart successfully compacted the virtual disk file.') {
                $finalSize = Get-ChildItem $Disk.FullName | Select-Object -ExpandProperty Length
                $finalSizeGB = [math]::Round( $finalSize / 1GB, 2 )
                $success = $true
                Remove-Item $tempFileName
            }
            else {
                Set-Content -Path "$env:TEMP\FslDiskPartError$($Disk.Name)-$retries.log" -Value $diskPartResult
                $retries++
                #if DiskPart fails, try, try again.
            }
            Start-Sleep 1
        }

        If ($success -ne $true) {
            Write-VhdOutput -DiskState "DiskShrinkFailed"
            Remove-Item $tempFileName
            return
        }

        #If I decide to add Hyper-V module support, I'll need this code later
        if ($hyperv -eq $true) {
            #Now we need to reinflate the partition to its previous size
            try {
                $mount = Mount-FslDisk -Path $Disk.FullName -PassThru
                $partInfo = Get-Partition -DiskNumber $mount.DiskNumber | Where-Object -Property 'Type' -EQ -Value 'Basic'
                Resize-Partition -InputObject $partInfo -Size $sizeMax -ErrorAction Stop
                $paramWriteVhdOutput = @{
                    DiskState    = "Success"
                    FinalSizeGB  = $finalSizeGB
                    SpaceSavedGB = $originalSizeGB - $finalSizeGB
                }
                Write-VhdOutput @paramWriteVhdOutput
            }
            catch {
                Write-VhdOutput -DiskState "PartitionSizeRestoreFailed"
                return
            }
            finally {
                $mount | DisMount-FslDisk
            }
        }

        $paramWriteVhdOutput = @{
            DiskState    = "Success"
            FinalSizeGB  = $finalSizeGB
            SpaceSavedGB = $originalSizeGB - $finalSizeGB
        }
        Write-VhdOutput @paramWriteVhdOutput
    } #Process
    END { } #End
}  #function Shrink-OneDisk